Quadriga/src/main/java/edu/asu/spring/quadriga/rest/TextUploadRestController.java

```diff
@@ -0,0 +1,125 @@
+     * @param response
+     *            Generic HTTP Response Object for returning the XML response
+     * @param request
+     *            Generic HTTP Request Object for RECEIVING the XML response
+     * @param xml
+     *            XML content as a string
+     * @return Returns a Response Entity that contains with the information
+     *         about the stored Textfile
+     * @throws RestException
+     */
+    @RequestMapping(value = "rest/project/{projectid}/workspace/{workspaceid}/uploadtext", method = RequestMethod.POST, produces = "application/xml")
+    public ResponseEntity<String> uploadText(@PathVariable("workspaceid") String wsId,
+            @PathVariable("projectid") String projId, HttpServletResponse response, HttpServletRequest request,
+            @RequestBody String xml) throws RestException {
+        String textURI = messages.getProperty("textfiles_location.url");
+        ITextFile txtFile = null;
+        try {
+            txtFile = txtXMLParser.parseTextXML(xml, wsId, projId);
+        } catch (TextFileParseException tfe) {
+            String errorMsg = errorMessageRest.getErrorMsg(tfe.getMessage());
+            return new ResponseEntity<String>(errorMsg, HttpStatus.BAD_REQUEST);
+        }
+        try {
+            tfManager.saveTextFile(txtFile);
+        } catch (QuadrigaStorageException e) {
+            String errorMsg = errorMessageRest.getErrorMsg(e.getMessage());
+            return new ResponseEntity<String>(errorMsg, HttpStatus.INTERNAL_SERVER_ERROR);
+        } catch (FileStorageException e) {
+            String errorMsg = errorMessageRest.getErrorMsg(e.getMessage());
+            return new ResponseEntity<String>(errorMsg, HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+
+        VelocityEngine engine = null;
+        try {
+            engine = restVelocityFactory.getVelocityEngine(request);
+            engine.init();
+            Template template = engine.getTemplate("velocitytemplates/textfile.vm");
+            VelocityContext context = new VelocityContext(restVelocityFactory.getVelocityContext());
+            context.put("textid", txtFile.getTextId());
+            context.put("refid", txtFile.getRefId());
+            context.put("filename", txtFile.getFileName());
+            context.put("wsid", txtFile.getWorkspaceId());
+            context.put("projid", txtFile.getProjectId());
+            context.put("texturi", textURI + "/" + txtFile.getTextId());
+            StringWriter writer = new StringWriter();
+            template.merge(context, writer);
+            HttpHeaders httpHeaders = new HttpHeaders();
+            httpHeaders.setContentType(MediaType.APPLICATION_XML);
+            return new ResponseEntity<String>(writer.toString(), httpHeaders, HttpStatus.CREATED);
+        } catch (FileNotFoundException e) {
+            throw new RestException(404, e);
+        } catch (IOException e) {
+            throw new RestException(404, e);
```

Comment:
404s should only be thrown when the client requests something the server can't "find" (e.g. request a project that doesn't exist). In this case the velocity engine can't find the template, right? So that would be a misconfiguration or error in installation and should be a 500.